### PR TITLE
mk: fix undefined variable warnings

### PR DESCRIPTION
**Pull request by tamird**
Some of this is scary stuff. Probably time to lint against this.

Found with `make --warn-undefined-variables`.

r? @alexcrichton

### CODE REVIEW

**Comment by tamird** [comment]
@alexcrichton this is ready for a look

File: configure [hunk 1]
```diff
@@ -679,9 +679,9 @@ case "$CFG_RELEASE_CHANNEL" in
 esac
 
 # Adjust perf and debug options for debug mode
-if [ -n "$CFG_ENABLE_DEBUG" ]; then
+if [ $CFG_ENABLE_DEBUG -eq 1 ]; then
```
Comment (alexcrichton) [inline configure]: How come these all changed? Do shells warn about this kind of thing?

File: configure [hunk 2]
```diff
@@ -260,14 +260,15 @@ opt_core() {
...
+        local V="CFG_${UFLAG}_${UOP}"
+        eval $V=0
         for arg in $CFG_CONFIGURE_ARGS
         do
             if [ "$arg" = "--${FLAG}-${OP}" ]
             then
-                OP=$(echo $OP | tr 'a-z-' 'A-Z_')
-                FLAG=$(echo $FLAG | tr 'a-z' 'A-Z')
-                local V="CFG_${FLAG}_${OP}"
-                local V_PROVIDED="CFG_${FLAG}_${OP}_PROVIDED"
```
Comment (alexcrichton) [inline configure]: Could you elaborate a bit on what's going on here? This seems unrelated to the makefile changes and our configure script is historically *very* tricky to modify

File: mk/docs.mk [hunk 3]
```diff
@@ -126,6 +126,7 @@ doc/not_found.html: $(D)/not_found.md $(HTML_DEPS) | doc/
 		--markdown-css https://doc.rust-lang.org/rust.css $<
 
 define DEF_DOC
+RUSTDOC_DEPS_$(1) ?=
```
Comment (alexcrichton) [inline mk/docs.mk]: I'm a little wary of sprinkling these around as they're often just intended to be empty. I'm not sure we gain a huge amount of benefit from ensuring that 100% of our variables are defined before use (e.g. did this uncover a bug?), so we may want to hold off on these for now.

Changes like `ifeq` to `ifdef` are of course fine though.

File: mk/platform.mk [hunk 4]
```diff
@@ -93,14 +93,14 @@ endif
 AR := ar
 
 define SET_FROM_CFG
+  $(1) ?=
```
Comment (alexcrichton) [inline mk/platform.mk]: e.g. this looks pretty worrisome to me...

File: configure (hunk 1 above)
Comment (tamird) [inline configure]: They don't warn; these had to change because of the addition of `eval $V = 0` on L266

File: configure (hunk 2 above)
Comment (tamird) [inline configure]: Yeah, it's an unrelated style change to bring this in line with `valopt_core()`. I initially did this with the hope of `save`ing zero-valued variables (because the makefile then reads those variables) but ended up not committing that bit. It seemed a shame to discard this extra strictness, so I kept it, albeit in a separate commit.

File: mk/docs.mk (hunk 3 above)
Comment (tamird) [inline mk/docs.mk]: These guys did not specifically uncover bugs, and I agree with you that sprinkling them around is gross. The value of adding these is that it suppresses some warnings; in a world where there are no spurious warnings, people are incented to investigate each warning they introduce.

That said, we are still a long way from that, so I'll back this out if you insist.

File: mk/platform.mk (hunk 4 above)
Comment (tamird) [inline mk/platform.mk]: How come? This is not different from the default behaviour, assuming that $(1) is used somewhere.

File: configure (hunk 2 above)
Comment (alexcrichton) [inline configure]: Hm ok, can you expand the commit message as to what's going on here? It's pretty unclear from "configure script cleanup" with a diff on some historically very-difficult-to-read code :(

Bringing things in line seems fine to me, however.

File: mk/docs.mk (hunk 3 above)
Comment (alexcrichton) [inline mk/docs.mk]: I think for now I'd prefer to avoid these changes. I'm worried about the sprinkling around and I suspect that different platforms will have different sets of warnings (encouraging more sprinklings), and there's nothing making sure that we actually have the "right set of sprinklings". In general this seems like it may end up just growing to be messy over time :(

File: configure (hunk 2 above)
Comment (tamird) [inline configure]: Done.

File: mk/docs.mk (hunk 3 above)
Comment (tamird) [inline mk/docs.mk]: Sounds reasonable. Removed.

**Comment by alexcrichton** [comment]
The travis failure here seems worrisome?

**Comment by tamird** [comment]
Yeah, I'll take a look tomorrow; out of time for the day!

**Comment by tamird** [comment]
@alexcrichton PTAL. I've taken out the `configure` changes which were troublesome.

**Comment by steveklabnik** [comment]
Ah wonderful! I had looked into trying to do this a while back, and didn't do it. Bravo! :+1: :100:

**Comment by steveklabnik** [comment]
There's also an open issue that this closes.... hm

**Comment by tamird** [comment]
@steveklabnik I wouldn't close that issue; this is only the tip of the iceberg :crying_cat_face:

**Comment by steveklabnik** [comment]
Ah, cool :) still good regardless.

**Comment by tamird** [comment]
I think the failures here are legit, looking into it

**Comment by tamird** [comment]
Well, I wasn't able to reproduce any failures locally. Can I get another r+?

**Comment by tamird** [comment]
@alexcrichton I believe I've fixed the android issue. Let's give it another go?

**Comment by tamird** [comment]
sigh, `ifeq` and `$(if (...)` are not interchangeable. once more, @alexcrichton? also, maybe it's ok to close this in light of #31123

**Comment by alexcrichton** [comment]
Ah we can see if this works and take it from there

@bors: r+ d03712977d7c913044f2b863269c4491d7fa7c36
